Prerender Issue — IOEXT-1811 (maxVersion error) (#711)
commerce-docs/microsite-commerce-storefront: src/content/docs/setup/configuration/aem-prerender.mdx

import Link from '@components/Link.astro';
import LinkCard from '@components/LinkCard.astro';
import TableWrapper from '@components/TableWrapper.astro';
import Options from '@components/Options.astro';
import Option from '@components/Option.astro';
import Tasks from '@components/Tasks.astro';
import Task from '@components/Task.astro';
import { Steps, Aside } from '@astrojs/starlight/components';

The AEM Commerce Prerender solution enables your Edge Delivery Services Commerce Storefront to serve data-hydrated product page markup to search engines and AI crawlers _before_ JavaScript executes. This can improve SEO rankings and makes your product catalog accessible to LLMs.

## How it works

> You deploy an App Builder application that runs continuously in Adobe's cloud, monitoring your catalog and generating complete HTML for each product. This HTML is published to Edge Delivery Services and served as the initial HTML "source" to page visitors. Then, your `product-details` block enhances the pre-rendered HTML with interactive features powered by the PDP drop-in. You should deploy the App Builder app first because your storefront integration depends on the HTML it generates.

Reviewer: ```suggestion
You deploy an App Builder application that runs continuously in Adobe’s cloud. This application monitors your catalog and generates fully rendered HTML for each product, which is published to Edge Delivery Services and served as the initial page source to visitors. On the client side, the product-details block then enhances this pre-rendered HTML with interactive capabilities provided by the PDP drop-in. Because the storefront integration relies on the generated HTML, you must deploy the App Builder application before configuring the storefront.
```

The <Link href="https://github.com/adobe-rnd/aem-commerce-prerender" text="AEM Commerce Prerender repository" /> provides complete installation instructions. The following overview highlights the key configuration steps and values you'll need.

<Tasks>

<Task>

### Clone the repository and install dependencies

```bash
git clone https://github.com/adobe-rnd/aem-commerce-prerender.git
cd aem-commerce-prerender
npm install
```

</Task>

<Task>

### Run the setup wizard

Run the setup wizard to generate your `.env` configuration:

```bash
npm run setup
```

> The wizard prompts for your App Builder workspace selection and Adobe I/O authentication.

Reviewer: ```suggestion
The wizard prompts you to select an App Builder workspace and authenticate with Adobe I/O.
```

</Task>

<Task>

### Configure environment variables

> Configure environment variables in the generated `.env` file with your storefront-specific values:

Reviewer: ```suggestion
Configure the environment variables in the generated `.env` file with values specific to your storefront.
```

| Variable | Description | Example |
|----------|-------------|---------|
| `ORG` | Your GitHub organization or username | `adobe` |
| `SITE` | Your AEM site/repository name | `your-storefront` |
| `CONTENT_URL` | Your AEM content URL (auto-populated by setup wizard) | `https://main--your-storefront--adobe.aem.live` |
| `STORE_URL` | Your Commerce store URL (auto-populated by setup wizard) | `https://main--your-storefront--adobe.aem.live` |
| `PRODUCTS_TEMPLATE` | URL for the product template page (auto-populated by setup wizard) | `https://main--your-storefront--adobe.aem.live/products/default` |
| `PRODUCT_PAGE_URL_FORMAT` | URL pattern for product pages (auto-populated by setup wizard). Supports tokens: `{locale}`, `{urlKey}`, `{sku}` | `/{locale}/products/{urlKey}` |
| `LOCALES` | Comma-separated list of locales (e.g., `en-us,en-gb,fr-fr`) or empty for non-localized sites | `en-us,fr-fr` |
| `AEM_ADMIN_API_AUTH_TOKEN` | Long-lived authentication token for AEM Admin API (valid for 1 year). The setup wizard exchanges your temporary token for this automatically. | `your-admin-token-here` |

</TableWrapper>

<Aside type="note" title="Commerce configuration">

> The prerender app reads Commerce configuration (endpoints, headers, and store context) from your storefront's `config.json` file, not from environment variables. Ensure your storefront has a properly configured `config.json` file with Commerce endpoints and headers as documented in the <Link href="/setup/configuration/commerce-configuration/" text="Storefront configuration" /> documentation. For Adobe Commerce Optimizer, configure the appropriate headers (`AC-View-ID`, `AC-Source-Locale`, `AC-Price-Book-ID`) in your storefront's `config.json` file.

Reviewer: ```suggestion
The prerender app reads Commerce configuration (endpoints, headers, and store context) from your storefront's `config.json` file, not from environment variables. Verify that your storefront has a properly configured `config.json` file with the required Commerce endpoints and headers as described in the <Link href="/setup/configuration/commerce-configuration/" text="Storefront configuration" /> documentation. For Adobe Commerce Optimizer, configure the appropriate headers (`AC-View-ID`, `AC-Source-Locale`, `AC-Price-Book-ID`) in your storefront's `config.json` file.
```

</Aside>

</Task>

<Task>

### Obtain required credentials

**AEM Admin API token**: To obtain your token:
1. Login to the <Link href="https://www.aem.live/docs/admin.html#setting-up-authentication-for-admin-api" text="AEM Admin API" /> (select the link from your preferred Identity Provider; the ones suffixed by "_sa" allow you to select a specific account rather than the one currently logged in)
2. Once redirected to the JSON response, open your browser's Developer Tools (F12)
3. Go to the Application tab (Chrome) or Storage tab (Firefox)
4. Under Cookies, find and copy the value of the `auth_token` cookie

> 5. Paste that token in the setup wizard textarea (the setup wizard exchanges your temporary token for a long-lived token automatically)

Reviewer: ```suggestion
**AEM Admin API token**: To obtain your token:
1. Login to the <Link href="https://www.aem.live/docs/admin.html#setting-up-authentication-for-admin-api" text="AEM Admin API" />  using your preferred Identity Provider.
   Providers with the "_sa" suffix allow you to select a specific account instead of using the one currently logged in.
2. After authentication, you are redirected to a page displaying a JSON response.
3. Open your browser's Developer Tools (F12).
4. Navigate to:
    - **Application** tab in Chrome
    - **Storage** tab (Firefox)
5. Under Cookies, locate the `auth_token` cookie and copy it's value.
6. Paste the token into the setup wizard textarea.

    The setup wizard exchanges your temporary token for a long-lived token automatically)
```

**App Builder credentials**: To configure the credentials:
1. Go to the <Link href="https://developer.adobe.com/console" text="Adobe Developer Console" />
2. Open your project and navigate to the workspace you want to use (Stage or Production)
3. Click the "Download All" button in the top-right to download the App Builder configuration file (a `.json` file)

> 4. Run `aio app use <file_name>` in the root directory of your prerender project to activate the credentials

Reviewer: ```suggestion
**App Builder credentials**: To configure the credentials:
1. Go to the <Link href="https://developer.adobe.com/console" text="Adobe Developer Console" />
2. Open your project and select the workspace you want to use (for example, Stage or Production)
3. In the top-right corner, click **Download All** to download the App Builder configuration file (a `.json` file)
4. In the root directory of your prerender project, run the following command to activate the credentials:
    ```bash
     aio app use <file_name>` 
    ```

> **Commerce credentials** (Adobe Commerce as a Cloud Service): Access your Commerce Admin → System → Services → <Link href="https://experienceleague.adobe.com/en/docs/commerce-merchant-services/user-guides/integration-services/saas" text="Commerce Services Connector" /> for environment and API configuration

Reviewer: ```suggestion
**Commerce credentials** (Adobe Commerce as a Cloud Service): For environment and API configuration, open your Commerce Admin and navigate to  **System → Services → <Link href="https://experienceleague.adobe.com/en/docs/commerce-merchant-services/user-guides/integration-services/saas" text="Commerce Services Connector" />**
```

> **Commerce credentials** (Adobe Commerce Optimizer): Configure your Commerce backend connection according to your deployment type. See the <Link href="/setup/configuration/commerce-configuration/" text="Storefront configuration" /> documentation for details on configuring Optimizer headers and endpoints.

Reviewer: ```suggestion
**Commerce credentials** (Adobe Commerce Optimizer): Configure your Commerce backend connection according to your deployment type. For details on configuring Optimizer headers and endpoints, see the <Link href="/setup/configuration/commerce-configuration/" text="Storefront configuration" /> documentation for details.
```

> **Note**: Commerce configuration (endpoints and headers) is read from your storefront's `config.json` file, not from environment variables. Ensure your storefront has a properly configured `config.json` file before deploying the prerender app.

Reviewer: ```suggestion
**Note**: The prerender app reads Commerce configuration (endpoints and headers) from your storefront's `config.json` file, not from environment variables. Before deployment, verify that `config.json` is correctly configured.
```

<Aside type="caution" title="Required: maxVersion in productDependencies">

> If your `app.config.yaml` includes a `productDependencies` array, **each entry must include `maxVersion`** or `aio app build` will fail with a validation error.

Reviewer: ```suggestion
If your `app.config.yaml` includes a `productDependencies` array, **each entry must include `maxVersion`**. Otherwise, the `aio app build` command fails with a validation error.
```

**Correct format:**
```yaml
productDependencies:
  - code: <PRODUCT_CODE>
    minVersion: <MIN_VERSION>
    maxVersion: <MAX_VERSION>  # Required
```

**Incorrect format (will cause build failure):**
```yaml
productDependencies:
  - code: <PRODUCT_CODE>
    minVersion: <MIN_VERSION>
    # Missing maxVersion - build will fail
```

See the Troubleshooting section below if you encounter this error.
</Aside>

</Task>

<Task>

### Deploy to App Builder

```bash
npm run deploy
```

> This deploys all actions and schedules to your selected workspace (Stage or Production).

Reviewer: ```suggestion
This command deploys all actions and schedules to your selected workspace (Stage or Production).
```

</Task>

<Task>

### Test the deployment

Manually invoke actions using the Management UI or via CLI commands:

```bash
# Fetch all products from Catalog Service and store them
aio rt action invoke aem-commerce-ssg/fetch-all-products

# Check for product changes and generate markup
aio rt action invoke aem-commerce-ssg/check-product-changes

# Clean up and unpublish deleted products
aio rt action invoke aem-commerce-ssg/mark-up-clean-up
```

> Verify product markup generation for sample SKUs and check that pages publish successfully to your AEM project.

Reviewer: ```suggestion
Verify that product markup generation for sample SKUs and that pages are published successfully to your AEM project.
```

</Task>

<Task>

### Enable automated synchronization

> 

Reviewer: ```suggestion

After deployment, perform the following steps:
```

- Use the <Link href="https://prerender.aem-storefront.com" text="Management UI" /> to start the change detector

> - Configure polling intervals in the `app.config.yaml` file (default: checks for changes every 5 minutes)

Reviewer: ```suggestion
- Configure polling intervals in the `app.config.yaml` file (by default, the app checks for changes every 5 minutes) 
```

| Variable | Description | Example |
|----------|-------------|---------|
| `ORG` | Your GitHub organization or username | `adobe` |
| `SITE` | Your AEM site/repository name | `your-storefront` |
| `CONTENT_URL` | Your AEM content URL (auto-populated by setup wizard) | `https://main--your-storefront--adobe.aem.live` |
| `STORE_URL` | Your Commerce store URL (auto-populated by setup wizard) | `https://main--your-storefront--adobe.aem.live` |
| `PRODUCTS_TEMPLATE` | URL for the product template page (auto-populated by setup wizard) | `https://main--your-storefront--adobe.aem.live/products/default` |
| `PRODUCT_PAGE_URL_FORMAT` | URL pattern for product pages (auto-populated by setup wizard). Supports tokens: `{locale}`, `{urlKey}`, `{sku}` | `/{locale}/products/{urlKey}` |
| `LOCALES` | Comma-separated list of locales (e.g., `en-us,en-gb,fr-fr`) or empty for non-localized sites | `en-us,fr-fr` |
| `AEM_ADMIN_API_AUTH_TOKEN` | Long-lived authentication token for AEM Admin API (valid for 1 year). The setup wizard exchanges your temporary token for this automatically. | `your-admin-token-here` |

</TableWrapper>

<Aside type="note" title="Commerce configuration">

> The prerender app reads Commerce configuration (endpoints, headers, and store context) from your storefront's `config.json` file, not from environment variables. Verify that your storefront has a properly configured `config.json` file with the required Commerce endpoints and headers as described in the <Link href="/setup/configuration/commerce-configuration/" text="Storefront configuration" /> documentation. For Adobe Commerce Optimizer, configure the appropriate headers (`AC-View-ID`, `AC-Source-Locale`, `AC-Price-Book-ID`) in your storefront's `config.json` file.

Reviewer 2: ```suggestion
The prerender app reads Commerce configuration (endpoints, headers, and store context) from your site's configuration, not from environment variables. Verify that your storefront has a properly configured site configuration with the required Commerce endpoints and headers as described in the <Link href="/setup/configuration/commerce-configuration/" text="Storefront configuration" /> documentation. For Adobe Commerce Optimizer, configure the appropriate headers (`AC-View-ID`, `AC-Source-Locale`) in your site configuration.
```

AC-Price-Book-ID should be automatic based on user auth state after next "release", assuming configuration set up correctly (adobe-commerce-optimizer: true).

Also let's stop using the config.json term. It should be "site configuration".

Reviewer 3: agree

**App Builder credentials**: To configure the credentials:
1. Go to the <Link href="https://developer.adobe.com/console" text="Adobe Developer Console" />
2. Open your project and select the workspace you want to use (for example, Stage or Production)
3. In the top-right corner, click **Download All** to download the App Builder configuration file (a `.json` file)
4. In the root directory of your prerender project, run the following command to activate the credentials:
    ```bash
     aio app use <file_name>` 
    ```

**Commerce credentials** (Adobe Commerce as a Cloud Service): For environment and API configuration, open your Commerce Admin and navigate to  **System → Services → <Link href="https://experienceleague.adobe.com/en/docs/commerce-merchant-services/user-guides/integration-services/saas" text="Commerce Services Connector" />**

**Commerce credentials** (Adobe Commerce Optimizer): Configure your Commerce backend connection according to your deployment type. For details on configuring Optimizer headers and endpoints, see the <Link href="/setup/configuration/commerce-configuration/" text="Storefront configuration" /> documentation for details.

> **Note**: The prerender app reads Commerce configuration (endpoints and headers) from your storefront's `config.json` file, not from environment variables. Before deployment, verify that `config.json` is correctly configured.

Reviewer 2: Same note re config.json vs "site configuration".

<Aside type="caution" title="Required: maxVersion in productDependencies">

> If your `app.config.yaml` includes a `productDependencies` array, **each entry must include `maxVersion`**. Otherwise, the `aio app build` command fails with a validation error.

Reviewer 2: Does the prerender repo come with a "default" or "reference" `app.config.yaml`? If so, should we just fix it there instead of making a docs change?

Or, is this a scenario that happens in specific situations, like maybe `productDependencies` is added if you have some specific version of `aio`?

Reviewer 3: There is a default template; all that's needed after testing is to uncomment the lines and deploy. I'm not familiar with maxVersion. We have never used it.

**Correct format:**
```yaml
productDependencies:
  - code: <PRODUCT_CODE>
    minVersion: <MIN_VERSION>
    maxVersion: <MAX_VERSION>  # Required
```

**Incorrect format (will cause build failure):**
```yaml
productDependencies:
  - code: <PRODUCT_CODE>
    minVersion: <MIN_VERSION>
    # Missing maxVersion - build will fail

> ```

Reviewer 3: I think it's not needed for this project

See the Troubleshooting section below if you encounter this error.
</Aside>

</Task>

<Task>

### Deploy to App Builder

```bash
npm run deploy
```

This command deploys all actions and schedules to your selected workspace (Stage or Production).

</Task>

<Task>

### Test the deployment

Manually invoke actions using the Management UI or via CLI commands:

```bash
# Fetch all products from Catalog Service and store them
aio rt action invoke aem-commerce-ssg/fetch-all-products

# Check for product changes and generate markup
aio rt action invoke aem-commerce-ssg/check-product-changes

# Clean up and unpublish deleted products
aio rt action invoke aem-commerce-ssg/mark-up-clean-up
```

Verify that product markup generation for sample SKUs and that pages are published successfully to your AEM project.

</Task>

<Task>

### Enable automated synchronization

After deployment, perform the following steps:

> - Use the <Link href="https://prerender.aem-storefront.com" text="Management UI" /> to start the change detector

Reviewer 3: The context will only appear here if we complete all the installation steps in the local environment.

For troubleshooting during development and operations, consult the <Link href="https://github.com/adobe-rnd/aem-commerce-prerender/blob/main/docs/RUNBOOK.md" text="prerender runbook" />.

<Options label="Issue" numbered={false}>

<Option>

### Build error: Missing maxVersion in productDependencies

**Error:** `must have required property 'maxVersion'` in `/productDependencies/0`

**Fix:** Add `maxVersion` to each entry in your `productDependencies` array:

```yaml
productDependencies:
  - code: <PRODUCT_CODE>
    minVersion: <MIN_VERSION>
    maxVersion: <MAX_VERSION>  # Required field
```

Then redeploy:

```bash
npm run deploy
```

<Aside type="note" title="productDependencies vs requiredProducts">

> The `productDependencies` field is validated by the Adobe I/O CLI schema. For publicly distributed apps, Adobe App Builder documentation uses `requiredProducts` instead. Both fields serve similar purposes but may be used in different contexts. If you're using `productDependencies`, ensure `maxVersion` is included for each entry to pass validation.

Reviewer: ```suggestion
The `productDependencies` field is validated against the Adobe I/O CLI schema. For publicly distributed apps, Adobe App Builder documentation references `requiredProducts` instead. Both fields serve similar purposes but can be used in different contexts. When using `productDependencies`, ensure that each entry  includes `maxVersion` to pass validation.
```

</Aside>

</Option>

<Option>

### Authentication failures

**Fix:** Generate a new AEM Admin API token and update your `.env` file:

1. Get a new token from the <Link href="https://www.aem.live/docs/admin.html#tag/orgConfig/operation/createOrgApiKey" text="AEM Admin API" />
2. Update `AEM_ADMIN_API_AUTH_TOKEN` in your `.env` file
3. Redeploy:

```bash
npm run deploy
```

**Verify token expiration:** Decode your token using <Link href="https://www.jwt.io/" text="jwt.io" /> or check your <Link href="https://www.aem.live/docs/admin.html#setting-up-authentication-for-admin-api" text="AEM project admin panel" />. Tokens expire after 1 year.

</Option>

<Option>

### Product not rendering

> **Fix:** Verify the product exists in Catalog Service and check action logs:

Reviewer: ```suggestion
**Fix:** Verify that the product exists in Catalog Service and check the action logs:
```

```bash
# Check action logs
aio rt activation list

# Query Catalog Service directly
# Use the GraphQL Product API: https://developer.adobe.com/commerce/services/graphql/catalog-service/products/
```

> **Verify:** Use the <Link href="https://prerender.aem-storefront.com" text="Management UI" /> to view the product index and confirm the SKU is listed.

Reviewer: ```suggestion
**Verify:** Use the <Link href="https://prerender.aem-storefront.com" text="Management UI" /> to view the product index and confirm that the SKU is listed.
```

For troubleshooting during development and operations, consult the <Link href="https://github.com/adobe-rnd/aem-commerce-prerender/blob/main/docs/RUNBOOK.md" text="prerender runbook" />.

<Options label="Issue" numbered={false}>

<Option>

> ### Build error: Missing maxVersion in productDependencies

Reviewer 3: Look at my comment above

**Error:** `must have required property 'maxVersion'` in `/productDependencies/0`

**Fix:** Add `maxVersion` to each entry in your `productDependencies` array:

```yaml
productDependencies:
  - code: <PRODUCT_CODE>
    minVersion: <MIN_VERSION>
    maxVersion: <MAX_VERSION>  # Required field

> ```

Reviewer 3: probably we should remove it
